code/game/machinery/computer/apc_control.dm

```diff
@@ -0,0 +1,208 @@
+					continue
+				if(filters["Charge Above"] && (APC.cell.charge / APC.cell.maxcharge) < filters["Charge Above"] / 100)
+					continue
+				if(filters["Charge Below"] && (APC.cell.charge / APC.cell.maxcharge) > filters["Charge Below"] / 100)
+					continue
+				if(filters["Responsive"] && !APC.aidisabled)
+					continue
+				dat += "<a href='?src=\ref[src];access_apc=\ref[APC]'>[A]</a><br>\
+				<b>Charge:</b> [APC.cell.charge] / [APC.cell.maxcharge] W ([round((APC.cell.charge / APC.cell.maxcharge) * 100)]%)<br>\
+				<b>Area:</b> [APC.area]<br>\
+				[APC.aidisabled || APC.panel_open ? "<font color='#FF0000'>APC does not respond to interface query.</font>" : "<font color='#00FF00'>APC responds to interface query.</font>"]<br><br>"
+			dat += "<a href='?src=\ref[src];check_logs=1'>Check Logs</a><br>"
+			dat += "<a href='?src=\ref[src];log_out=1'>Log Out</a><br>"
+			if(emagged)
+				dat += "<font color='#FF0000'>WARNING: Logging functionality partially disabled from outside source.</font><br>"
+				dat += "<a href='?src=\ref[src];restore_logging=1'>Restore logging functionality?</a><br>"
+		else
+			if(logs.len)
+				for(var/entry in logs)
+					dat += "[entry]<br>"
+			else
+				dat += "<i>No activity has been recorded at this time.</i><br>"
+			if(emagged)
+				dat += "<a href='?src=\ref[src];clear_logs=1'><font color='#FF0000'>@#%! CLEAR LOGS</a>"
+			dat += "<a href='?src=\ref[src];check_apcs=1'>Return</a>"
+		operator = user
+	else
+		dat = "<a href='?src=\ref[src];authenticate=1'>Please swipe a valid ID to log in...</a>"
+	var/datum/browser/popup = new(user, "apc_control", name, 600, 400)
+	popup.set_content(dat)
+	popup.set_title_image(user.browse_rsc_icon(icon, icon_state))
+	popup.open()
+
+/obj/machinery/computer/apc_control/Topic(href, href_list)
+	if(..())
+		return
+	var/image/I = image(src) //For feedback message flavor
+	if(href_list["authenticate"])
+		var/obj/item/weapon/card/id/ID = usr.get_active_held_item()
+		if(!istype(ID))
+			ID = usr.get_idcard()
+		if(ID && istype(ID))
+			if(check_access(ID))
+				authenticated = TRUE
+				auth_id = "[ID.registered_name] ([ID.assignment])"
+				log_activity("logged in")
+		if(!authenticated) //Check for emags
+			var/obj/item/weapon/card/emag/E = usr.get_active_held_item()
+			if(E && istype(E))
+				usr << "<span class='warning'>You bypass [src]'s access requirements using your emag.</span>"
+				authenticated = TRUE
+				log_activity("logged in") //Auth ID doesn't change, hinting that it was illicit
+	if(href_list["log_out"])
+		log_activity("logged out")
+		authenticated = FALSE
+		auth_id = "\[NULL\]"
+	if(href_list["restore_logging"])
+		to_chat(usr, "<span class='robot notice'>\icon[I] Logging functionality restored from backup data.</span>")
+		emagged = FALSE
+		LAZYADD(logs, "<b>-=- Logging restored to full functionality at this point -=-</b>")
+	if(href_list["access_apc"])
```

Comment:
you need to recheck all assumptions in the topic call, distance fo player, player status, if the apc is valid to intefere with etc

cannot rely on byond to only provide valid topic calls.